metal/atomic.h

```diff
@@ -0,0 +1,299 @@
+/* Copyright 2019 SiFive, Inc */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+#ifndef METAL__ATOMIC_H
+#define METAL__ATOMIC_H
+
+#include <stdint.h>
+
+#include <metal/compiler.h>
+
+typedef int32_t metal_atomic_t;
+
+#define METAL_ATOMIC_DECLARE(name)                                             \
+    __attribute((section(".data.atomics"))) metal_atomic_t name
+
+#define _METAL_STORE_AMO_ACCESS_FAULT 7
+
+/*!
+ * @brief Check if the platform supports atomic operations
+ *
+ * @return 1 if atomic operations are supported, 0 if not
+ */
+__inline__ int32_t metal_atomic_available(void) {
+#ifdef __riscv_atomic
+    return 1;
+#else
+    return 0;
+#endif
+}
+
+/*!
+ * @brief Atomically increment a metal_atomic_t and return its old value
+ *
+ * If atomics are not supported on the platform, this function will trap with
+ * a Store/AMO access fault.
+ *
+ * @param a The pointer to the value to increment
+ * @param increment the amount to increment the value
+ *
+ * @return The previous value of the metal_atomic_t
+ */
+__inline__ int32_t metal_atomic_add(metal_atomic_t *a, int32_t increment) {
+#ifdef __riscv_atomic
+    int32_t old;
+    __asm__ volatile("amoadd.w %[old], %[increment], (%[atomic])"
+                     : [old] "=r"(old)
+                     : [increment] "r"(increment), [atomic] "r"(a)
+                     : "memory");
+    return old;
+#else
+    /* Store the memory address in mtval like a normal store/amo access fault */
+    __asm__("csrw mtval, %[atomic]" ::[atomic] "r"(a);
+
+    _metal_trap(_METAL_STORE_AMO_ACCESS_FAULT);
+
+    /* If execution returns, we have to return something, so just pick 0 */
+    return 0;
+#endif
+}
+
+/*!
+ * @brief Atomically bitwise-AND a metal_atomic_t and return its old value
+ *
+ * If atomics are not supported on the platform, this function will trap with
+ * a Store/AMO access fault.
+ *
+ * @param a The pointer to the value to bitwise-AND
+ * @param mask the bitmask to AND
+ *
+ * @return The previous value of the metal_atomic_t
+ */
+__inline__ int32_t metal_atomic_and(metal_atomic_t *a, int32_t mask) {
+#ifdef __riscv_atomic
+    int32_t old;
+    __asm__ volatile("amoand.w %[old], %[mask], (%[atomic])"
+                     : [old] "=r"(old)
+                     : [mask] "r"(mask), [atomic] "r"(a)
+                     : "memory");
+    return old;
+#else
+    /* Store the memory address in mtval like a normal store/amo access fault */
```

Comment:
Since these amo access fault handling are the same, having a macro avoid multiple places to change